File: Strohm/android/strohm-native/src/main/java/dev/strohmnative/Log.kt
package dev.strohmnative

class Log {
    companion object {
        fun debug(vararg args: Any) {
            log(Level.DEBUG, args)
        }

        fun info(vararg args: Any) {
            log(Level.INFO, args)
        }

        fun warn(vararg args: Any) {
            log(Level.WARN, args)
        }

        fun error(vararg args: Any) {
            log(Level.ERROR, args)
        }

        private fun log(level: Level, args: Array<out Any>) {
            val argsWithLevel = arrayListOf<Any>(level.rawValue).plus(args)
            val encodedArgs = StrohmNative.getInstance().comms.encode(argsWithLevel)

            StrohmNative.getInstance().call("globalThis.strohm_native.log.log_from_native(\"$encodedArgs\")")
        }
    }

    enum class Level(val rawValue: String) {
        DEBUG("debug"),
        INFO("info"),
        WARN("warn"),
        ERROR("error")
    }
}

File: Strohm/ios/StrohmNative/Sources/StrohmNative/Log.swift
import Foundation

class Log {
    static func debug(_ args: Any...) {
        Log.log(level: .debug, args)
    }

    static func info(_ args: Any...) {
        Log.log(level: .info, args)
    }

    static func warn(_ args: Any...) {
        Log.log(level: .warn, args)
    }

    static func error(_ args: Any...) {
        Log.log(level: .error, args)
    }

    static private func log(level: Level, _ args: [Any]) {
        print("[\(level)]", args)

        let argsWithLevel: [Any] = ["\(level)"] + args
        guard let encodedArgs = StrohmNative.default.comms.encode(object: argsWithLevel) else {
            print("[error] Failed to encode log args", argsWithLevel)
            return
        }

        let encoded = encodedArgs.replacingOccurrences(of: "\"", with: "\\\"")
        StrohmNative.default.call(method: "globalThis.strohm_native.log.log_from_native(\"\(encoded)\")") { cljsError in
            print("[fatal] Call to log error in cljs failed with error \(cljsError)")
        }
    }

    enum Level: String {
        case debug
        case info
        case warn
        case error
    }
}
